add more optional module locations

underdoeg opened the pull request:
Introduces the macro ```add_island_module_location``` All potential module directories are stored in a list which can be extended. 

One use case for this is to have additional modules that are not part of the official island eco system but can still be checked into a git repo together with the app.

Locations added with add_island_module_location take precedence over the default modules directory. This allows modules to be overriden on a per project basis if needed.

```
add_island_module_location(${CMAKE_CURRENT_SOURCE_DIR}/../modules)
```

underdoeg committed:
add more optional module locations

tgfrerer commented:
Ohh this looks really great! I've wanted to have something like that for a good while :)

I'd like to give this a good run-through on all currently supported platforms and compilers, and test it with Debug and Release targets. I'd also like to test how it interacts with the current option in the modules CMakeLists.txt files to specify module dependencies. 

I think I can spend some time tomorrow morning doing these tests...

underdoeg commented:
Great go for it. I am fairly confident that it should be fine, since the implementation is very non intrusive.
But half the time I say something like that it breaks regardless...

underdoeg committed:
fix include path for third party modules

underdoeg commented:
I didn't manage to automatically add third party module include paths to the app module, so I had to update the templates to include them. Older projects should not be affected by this though. (as long as they do not try to add a module outside of the default path)

tgfrerer committed:
[CMakeLists.txt] add missing forward slash

Add missing directory separator after ${ISLAND_BASE_DIR} in app
CMakeLists.txt files and templates.

This is so that we can follow conventions by specifying the Island base
dir without having to add a trailing '/' character.

tgfrerer committed:
[whitespace] reformat / fix typos

tgfrerer committed:
[CMakeLists] move module include_directories higher

optional module locations must be exported **before** any modules
(including the main application module) get loaded - therefore we don't
defer adding any optional module locations to our include directories.

tgfrerer committed:
[CMakeLists] add doc comment

tgfrerer committed:
[CMakeLists.txt] add default search path for modules

use the default mechanism for setting a module search path to set the
default module search path this will allow us to override things if need
be, in the future.

tgfrerer committed:
[CMakeLists] update all examples and templates

update CmakeLists.txt for all examples and templates so that we're
consistent with the sequencing of things

tgfrerer commented:
I've tested on Linux and had to add a few minor tweaks so that out-of-source builds with optional modules would still build in Relase, where the linking order becomes important and a bit finnicky.

 Let's see if CI passes which will also tell us something about the state of things on Windows.
Still have to move over to Windows and test an actual out-of-tree build, but should CI pass, I'd be fairly confident that things will work.

Linux 
- [x] test out-of-tree build
- [x] pass CI

Windows Checklist: 
- [x] pass CI 
- [x] test out-of-tree build
- [x] test hot-reloading
- [x] test hot-reloading with out-of-tree build

underdoeg commented:
why did you ever add windows support? ;) no all good, let's hope both platforms behave the same

tgfrerer commented:
All tests passed, and windows out-of-tree compilation, and hot-reloading confirmed working. windows release builds also confirmed working, in-and out-of tree.

Merged - Thanks a lot!!

underdoeg commented:
great. this is useful
